# Data science in Python

- Course GitHub repo: https://github.com/pycam/python-data-science
- Python website: https://www.python.org/ 

## Session 2.3 live: Biological data with BioPython

**REMIND** participants to use red/pink post-it if they are lost during the live coding session to get help for the other trainer. 

In [1]:
import pandas 
mouse = pandas.read_csv('data/GRCm38.gff3', sep = '\t')

In [2]:
print(mouse.columns)

Index(['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase',
       'attributes', 'gbid'],
      dtype='object')


In [3]:
mouse[mouse['type']=='exon']

seqid      source  type      start        end score strand phase  \
2      NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   
3      NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   
7      NC_000073.6      Gnomon  exon  118329486  118329612     .      +     .   
8      NC_000076.6      Gnomon  exon    5175722    5175928     .      -     .   
9      NC_000083.6      Gnomon  exon   24585923   24586066     .      +     .   
13     NC_000077.6      Gnomon  exon   61740946   61741001     .      -     .   
14     NC_000070.6  BestRefSeq  exon  149130153  149130219     .      -     .   
15     NC_000068.7  BestRefSeq  exon  125218141  125218252     .      +     .   
16     NC_000086.7  BestRefSeq  exon  134446962  134447027     .      +     .   
18     NC_000076.6      Gnomon  exon   57708192   57708290     .      +     .   
19     NC_000077.6  BestRefSeq  exon   60499325   60499560     .      +     .   
21     NC_000067.6      Gnomon  exon  154426526  154426617     .      -     .   
23     NC_000081.6  BestRefSeq  exon   77153530   77153812     .      -     .   
24     NC_000069.6      Gnomon  exon   80702458   80702656     .      -     .   
25     NC_000071.6      Gnomon  exon   52622299   52622981     .      -     .   
26     NC_000075.6      Gnomon  exon   49502134   49505563     .      -     .   
28     NC_000085.6      Gnomon  exon   56842212   56842319     .      +     .   
32     NC_000081.6      Gnomon  exon   76380468   76380576     .      +     .   
34     NC_000073.6      Gnomon  exon   45309098   45309230     .      -     .   
35     NC_000076.6      Gnomon  exon  109903062  109903114     .      -     .   
38     NC_000085.6  BestRefSeq  exon    6372301    6372416     .      +     .   
39     NC_000068.7  BestRefSeq  exon  152440765  152440922     .      +     .   
41     NC_000074.6      Gnomon  exon   84923731   84923939     .      -     .   
43     NC_000071.6  BestRefSeq  exon   67702980   67703044     .      -     .   
45     NC_000071.6      Gnomon  exon  123646083  123646305     .      -     .   
46     NC_000083.6      Gnomon  exon   12895512   12895597     .      -     .   
48     NC_000086.7      Gnomon  exon  100612052  100612111     .      +     .   
49     NC_000074.6  BestRefSeq  exon   94746902   94747025     .      +     .   
52     NC_000079.6      Gnomon  exon   53485797   53485922     .      +     .   
54     NC_000074.6  BestRefSeq  exon  111011285  111011462     .      -     .   
...            ...         ...   ...        ...        ...   ...    ...   ...   
29952  NC_000074.6      Gnomon  exon  127371496  127371878     .      +     .   
29953  NC_000075.6  BestRefSeq  exon   21835510   21835814     .      +     .   
29956  NC_000075.6      Gnomon  exon   42597854   42597961     .      -     .   
29958  NC_000077.6      Gnomon  exon   34138269   34139431     .      +     .   
29959  NC_000085.6      Gnomon  exon   28928548   28928632     .      -     .   
29960  NC_000067.6      Gnomon  exon   34282423   34282592     .      +     .   
29961  NC_000077.6      Gnomon  exon   90600186   90600280     .      -     .   
29962  NC_000071.6      Gnomon  exon  101863777  101863896     .      -     .   
29963  NC_000067.6      Gnomon  exon  118545393  118545456     .      +     .   
29964  NC_000070.6  BestRefSeq  exon   46545612   46545728     .      -     .   
29966  NC_000070.6      Gnomon  exon   82970635   82970820     .      -     .   
29967  NC_000082.6      Gnomon  exon   43072424   43072469     .      +     .   
29970  NC_000073.6  BestRefSeq  exon   16205397   16205498     .      -     .   
29972  NC_000076.6      Gnomon  exon   52075813   52075975     .      -     .   
29973  NC_000073.6  BestRefSeq  exon  116154633  116154686     .      -     .   
29974  NC_000077.6      Gnomon  exon   87737551   87737668     .      +     .   
29976  NC_000082.6  BestRefSeq  exon   16316334   16316443     .      -     .   
29978  NC_000080.6      Gnomon  exon   65174

In [4]:
mouse.head()

seqid      source  type      start        end score strand phase  \
0  NC_000073.6      Gnomon   CDS   49420174   49420340     .      +     1   
1  NC_000067.6      Gnomon   CDS   64607740   64608137     .      -     2   
2  NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   
3  NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   
4  NC_000068.7      Gnomon   CDS  135932249  135932302     .      +     1   

                                          attributes            gbid  
0  ID=cds30916;Parent=rna43633;Dbxref=GeneID:7828...  XP_006541352.1  
1  ID=cds1286;Parent=rna1815;Dbxref=GeneID:67099,...  XP_006496267.1  
2  ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...     NM_177676.6  
3  ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...  XM_017312883.1  
4  ID=cds9263;Parent=rna13273;Dbxref=GeneID:18798...  XP_006499009.1

In [5]:
mouse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29998 entries, 0 to 29997
Data columns (total 10 columns):
seqid         29998 non-null object
source        29998 non-null object
type          29998 non-null object
start         29998 non-null int64
end           29998 non-null int64
score         29998 non-null object
strand        29998 non-null object
phase         29998 non-null object
attributes    29998 non-null object
gbid          28741 non-null object
dtypes: int64(2), object(8)
memory usage: 2.3+ MB


In [6]:
mouse.isna().sum()

seqid            0
source           0
type             0
start            0
end              0
score            0
strand           0
phase            0
attributes       0
gbid          1257
dtype: int64

In [7]:
mouse.dropna(inplace = True)

In [8]:
mouse.isna().sum()

seqid         0
source        0
type          0
start         0
end           0
score         0
strand        0
phase         0
attributes    0
gbid          0
dtype: int64

In [9]:
mouse.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 28741 entries, 0 to 29997
Data columns (total 10 columns):
seqid         28741 non-null object
source        28741 non-null object
type          28741 non-null object
start         28741 non-null int64
end           28741 non-null int64
score         28741 non-null object
strand        28741 non-null object
phase         28741 non-null object
attributes    28741 non-null object
gbid          28741 non-null object
dtypes: int64(2), object(8)
memory usage: 2.4+ MB


In [10]:
exon_mouse = mouse[mouse['type']=='exon']

In [11]:
exon_mouse

seqid      source  type      start        end score strand phase  \
2      NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   
3      NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   
7      NC_000073.6      Gnomon  exon  118329486  118329612     .      +     .   
8      NC_000076.6      Gnomon  exon    5175722    5175928     .      -     .   
9      NC_000083.6      Gnomon  exon   24585923   24586066     .      +     .   
13     NC_000077.6      Gnomon  exon   61740946   61741001     .      -     .   
14     NC_000070.6  BestRefSeq  exon  149130153  149130219     .      -     .   
15     NC_000068.7  BestRefSeq  exon  125218141  125218252     .      +     .   
16     NC_000086.7  BestRefSeq  exon  134446962  134447027     .      +     .   
18     NC_000076.6      Gnomon  exon   57708192   57708290     .      +     .   
19     NC_000077.6  BestRefSeq  exon   60499325   60499560     .      +     .   
21     NC_000067.6      Gnomon  exon  154426526  154426617     .      -     .   
23     NC_000081.6  BestRefSeq  exon   77153530   77153812     .      -     .   
24     NC_000069.6      Gnomon  exon   80702458   80702656     .      -     .   
25     NC_000071.6      Gnomon  exon   52622299   52622981     .      -     .   
26     NC_000075.6      Gnomon  exon   49502134   49505563     .      -     .   
28     NC_000085.6      Gnomon  exon   56842212   56842319     .      +     .   
32     NC_000081.6      Gnomon  exon   76380468   76380576     .      +     .   
34     NC_000073.6      Gnomon  exon   45309098   45309230     .      -     .   
35     NC_000076.6      Gnomon  exon  109903062  109903114     .      -     .   
38     NC_000085.6  BestRefSeq  exon    6372301    6372416     .      +     .   
39     NC_000068.7  BestRefSeq  exon  152440765  152440922     .      +     .   
41     NC_000074.6      Gnomon  exon   84923731   84923939     .      -     .   
43     NC_000071.6  BestRefSeq  exon   67702980   67703044     .      -     .   
45     NC_000071.6      Gnomon  exon  123646083  123646305     .      -     .   
46     NC_000083.6      Gnomon  exon   12895512   12895597     .      -     .   
48     NC_000086.7      Gnomon  exon  100612052  100612111     .      +     .   
49     NC_000074.6  BestRefSeq  exon   94746902   94747025     .      +     .   
52     NC_000079.6      Gnomon  exon   53485797   53485922     .      +     .   
54     NC_000074.6  BestRefSeq  exon  111011285  111011462     .      -     .   
...            ...         ...   ...        ...        ...   ...    ...   ...   
29952  NC_000074.6      Gnomon  exon  127371496  127371878     .      +     .   
29953  NC_000075.6  BestRefSeq  exon   21835510   21835814     .      +     .   
29956  NC_000075.6      Gnomon  exon   42597854   42597961     .      -     .   
29958  NC_000077.6      Gnomon  exon   34138269   34139431     .      +     .   
29959  NC_000085.6      Gnomon  exon   28928548   28928632     .      -     .   
29960  NC_000067.6      Gnomon  exon   34282423   34282592     .      +     .   
29961  NC_000077.6      Gnomon  exon   90600186   90600280     .      -     .   
29962  NC_000071.6      Gnomon  exon  101863777  101863896     .      -     .   
29963  NC_000067.6      Gnomon  exon  118545393  118545456     .      +     .   
29964  NC_000070.6  BestRefSeq  exon   46545612   46545728     .      -     .   
29966  NC_000070.6      Gnomon  exon   82970635   82970820     .      -     .   
29967  NC_000082.6      Gnomon  exon   43072424   43072469     .      +     .   
29970  NC_000073.6  BestRefSeq  exon   16205397   16205498     .      -     .   
29972  NC_000076.6      Gnomon  exon   52075813   52075975     .      -     .   
29973  NC_000073.6  BestRefSeq  exon  116154633  116154686     .      -     .   
29974  NC_000077.6      Gnomon  exon   87737551   87737668     .      +     .   
29976  NC_000082.6  BestRefSeq  exon   16316334   16316443     .      -     .   
29978  NC_000080.6      Gnomon  exon   65174

In [12]:
exon_mouse['gbid'].iloc[0]

'NM_177676.6'

In [13]:
from Bio import Entrez
from Bio import SeqIO
from Bio.SeqUtils import GC, molecular_weight

Entrez.email = 'Hello@example.com'
handle = Entrez.efetch(db = 'nucleotide', id = exon_mouse["gbid"].iloc[0], rettype = 'gb')
seq_record = SeqIO.read(handle, 'gb')
print(molecular_weight(seq_record.seq))
handle.close()

829310.7034000154


In [14]:
exon_mouse10 = exon_mouse.iloc[0:10,:]

In [15]:
exon_mouse10.reset_index(drop=True, inplace=True)

In [16]:
mws = []
for gbid in exon_mouse10["gbid"]:
    handle = Entrez.efetch(db = 'nucleotide', id = gbid, rettype = 'gb')
    seq_record = SeqIO.read(handle, 'gb')
    mws.append(molecular_weight(seq_record.seq))
    handle.close()

exon_mouse10.insert(len(exon_mouse10.columns), 'mw', mws)

In [17]:
exon_mouse10

seqid      source  type      start        end score strand phase  \
0  NC_000071.6  BestRefSeq  exon   24544129   24544373     .      -     .   
1  NC_000074.6      Gnomon  exon   71655663   71655741     .      -     .   
2  NC_000073.6      Gnomon  exon  118329486  118329612     .      +     .   
3  NC_000076.6      Gnomon  exon    5175722    5175928     .      -     .   
4  NC_000083.6      Gnomon  exon   24585923   24586066     .      +     .   
5  NC_000077.6      Gnomon  exon   61740946   61741001     .      -     .   
6  NC_000070.6  BestRefSeq  exon  149130153  149130219     .      -     .   
7  NC_000068.7  BestRefSeq  exon  125218141  125218252     .      +     .   
8  NC_000086.7  BestRefSeq  exon  134446962  134447027     .      +     .   
9  NC_000076.6      Gnomon  exon   57708192   57708290     .      +     .   

                                          attributes            gbid  \
0  ID=id325971;Parent=rna28488;Dbxref=GeneID:2310...     NM_177676.6   
1  ID=id568467;Parent=rna50916;Dbxref=GeneID:3820...  XM_017312883.1   
2  ID=id520282;Parent=rna46416;Dbxref=GeneID:1026...     XR_378454.2   
3  ID=id673061;Parent=rna60010;Dbxref=GeneID:6400...  XM_017314047.1   
4  ID=id1054336;Parent=rna94586;Dbxref=GeneID:187...  XM_017317334.1   
5  ID=id756287;Parent=rna67525;Dbxref=GeneID:2126...  XM_006532761.3   
6  ID=id309255;Parent=rna27107;Dbxref=GeneID:6992...     NM_027263.2   
7  ID=id157624;Parent=rna12621;Dbxref=GeneID:2049...  NM_001079690.1   
8  ID=id1188223;Parent=rna107372;Dbxref=GeneID:13...     NM_010078.3   
9  ID=id686725;Parent=rna61293;Dbxref=GeneID:1876...  XM_006512605.3   

             mw  
0  8.293107e+05  
1  2.164587e+06  
2  3.169640e+05  
3  8.345749e+06  
4  4.384230e+06  
5  5.642921e+05  
6  2.891971e+05  
7  1.441759e+06  
8  2.157642e+06  
9  3.328927e+05

**NOTE** We have to filter on exons to be able to query the Entrez nucleotide database otherwise there will be gbid for other types and you will need to specify another db value e.g. protein.

## Next session

Go to our next notebook: [Session 2.4: Data project report in Jupyter](24_python_data.ipynb)